src/AppInstallerCLICore/Commands/UpdateCommand.cpp

```diff
@@ -0,0 +1,105 @@
+#include "Workflows/WorkflowBase.h"
+#include "Resources.h"
+
+using namespace AppInstaller::CLI::Execution;
+using namespace AppInstaller::Manifest;
+using namespace AppInstaller::CLI::Workflow;
+
+namespace AppInstaller::CLI
+{
+    std::vector<Argument> UpdateCommand::GetArguments() const
+    {
+        return {
+            Argument::ForType(Args::Type::Query),
+            Argument::ForType(Args::Type::Manifest),
+            Argument::ForType(Args::Type::Id),
+            Argument::ForType(Args::Type::Name),
+            Argument::ForType(Args::Type::Moniker),
+            Argument::ForType(Args::Type::Version),
+            Argument::ForType(Args::Type::Channel),
+            Argument::ForType(Args::Type::Source),
+            Argument::ForType(Args::Type::Exact),
+            Argument::ForType(Args::Type::Interactive),
+            Argument::ForType(Args::Type::Silent),
+            Argument::ForType(Args::Type::Language),
+            Argument::ForType(Args::Type::Log),
+            Argument::ForType(Args::Type::Override),
+            Argument::ForType(Args::Type::InstallLocation),
+            Argument{ "force", Argument::NoAlias, Args::Type::Force, Resource::String::InstallForceArgumentDescription, ArgumentType::Flag },
+            Argument{ "all", Argument::NoAlias, Args::Type::All, Resource::String::UpdateAllArgumentDescription, ArgumentType::Flag },
+        };
+    }
+
+    Resource::LocString UpdateCommand::ShortDescription() const
+    {
+        return { Resource::String::UpdateCommandShortDescription };
+    }
+
+    Resource::LocString UpdateCommand::LongDescription() const
+    {
+        return { Resource::String::UpdateCommandLongDescription };
+    }
+
+    void UpdateCommand::Complete(Execution::Context& context, Execution::Args::Type valueType) const
+    {
+        switch (valueType)
+        {
+        case Execution::Args::Type::Query:
+        case Execution::Args::Type::Manifest:
+        case Execution::Args::Type::Id:
+        case Execution::Args::Type::Name:
+        case Execution::Args::Type::Moniker:
+        case Execution::Args::Type::Version:
+        case Execution::Args::Type::Channel:
+        case Execution::Args::Type::Source:
+            context <<
+                Workflow::CompleteWithSingleSemanticsForValue(valueType);
+            break;
+        case Execution::Args::Type::Language:
+            // May well move to CompleteWithSingleSemanticsForValue,
+            // but for now output nothing.
+            context <<
+                Workflow::CompleteWithEmptySet;
+            break;
+        case Execution::Args::Type::Log:
+            // Intentionally output nothing to allow pass through to filesystem.
+            break;
+        }
+    }
+
+    std::string UpdateCommand::HelpLink() const
+    {
+        // TODO: point to correct location
+        return "https://aka.ms/winget-command-update";
+    }
+
+    void UpdateCommand::ExecuteInternal(Execution::Context& context) const

```

Comment:
>ExecuteInternal [](start = 24, length = 15)

There should be three distinct scenarios for update per the spec:
1. `update` :: Basically the same as `list` but filtered to only packages with updates available.
2. `update foo` :: Searches for package `foo`, proceed to update if found.
3. `update --all` :: Equivalent to using the output from 1 to run 2 for each package.

Reply:
And in all cases I think it would be very beneficial to have the common steps of the workflow (source set up, search) performed once, and only diverge once needed.  As you have it here, every path does its own setup, and the nested workflows make it very hard to follow.

---
In reply to: [499918627](https://github.com/microsoft/winget-cli/pull/588#discussion_r499918627) [](ancestors = 499918627)